Improvements to Process List page

djmattyg007 opened the pull request:
- Add auto-refresh capability
- Add ability to show full queries in table

djmattyg007 commented:
I've rebased this pull request on top of the latest work on the master branch.

djmattyg007 commented:
I have again rebased this pull request on top of the latest work on the master branch. I've also added another commit that stops the autorefresh from occurring if an error occurs, such as the user being logged out. This caused issues if (for example) a MySQL server was shut down while the autorefresh is active, as it would keep autorefreshing trying to log the user back in, and eventually hit the 30-minute lockout.

vrana commented on adminer/include/design.inc.php:
```diff
@@ -22,6 +22,9 @@ function page_header($title, $error = "", $breadcrumb = array(), $title2 = "") {
 <meta http-equiv="Content-Script-Type" content="text/javascript">
 <meta name="robots" content="noindex">
 <meta name="referrer" content="origin-when-crossorigin">
+<?php if (!$error && isset($_GET["processlist"], $_GET["autorefresh"])) { ?>
```
Move to a `header` call in `processlist.inc.php`.

djmattyg007 replied:
I'm not sure what you mean by this sorry. Are you saying that I should replace the `meta http-equiv` tag with a `Location` header?

vrana replied:
`header("Refresh: " . intval($_GET["autorefresh"]))`

vrana commented on adminer/processlist.inc.php:
```diff
@@ -13,7 +13,7 @@
 
 <form action="" method="post">
-<table cellspacing="0" onclick="tableClick(event);" ondblclick="tableClick(event, true);" class="nowrap checkable">
+<table cellspacing="0" onclick="tableClick(event);" ondblclick="tableClick(event, true);" class="<?php if (!isset($_GET["showfull"])):?>nowrap <?php endif; ?>checkable">
```
Use `{}` instead of `endif`.

vrana commented on adminer/processlist.inc.php:
```diff
@@ -36,7 +36,7 @@
 			($jush == "pgsql" && $key == "current_query" && $val != "<IDLE>") ||
 			($jush == "oracle" && $key == "sql_text" && $val != "")
-			? "<code class='jush-$jush'>" . shorten_utf8($val, 100, "</code>") . ' <a href="' . h(ME . ($row["db"] != "" ? "db=" . urlencode($row["db"]) . "&" : "") . "sql=" . urlencode($val)) . '">' . lang('Clone') . '</a>'
+			? "<code class='jush-$jush'>" . (isset($_GET["showfull"]) ? $val . "</code>" : shorten_utf8($val, 100, "</code>")) . ' <a href="' . h(ME . ($row["db"] != "" ? "db=" . urlencode($row["db"]) . "&" : "") . "sql=" . urlencode($val)) . '">' . lang('Clone') . '</a>'
```
Use `h($val)`.

vrana commented on adminer/processlist.inc.php:
```diff
@@ -51,6 +51,24 @@
 	echo "<p><input type='submit' value='" . lang('Kill') . "'>\n";
 }
+if (isset($_GET["showfull"])) {
+	echo "<p><a href='" . h(ME . "processlist=") . "'>" . lang('Hide full') . "</a>";
```
Add these messages to `lang/xx.inc.php`.

vrana commented on adminer/processlist.inc.php:
```diff
@@ -51,6 +51,24 @@
+<?php if (isset($_GET["autorefresh"])) { ?>
+<a href="<?php echo remove_from_uri("autorefresh"); ?>"><?php echo lang('Stop auto refresh'); ?></a>
+<?php } else { ?>
+<form action="<?php echo ME; ?>" method="get">
```
Delete `<?php echo ME; ?>` (it's the default). `ME` should be wrapped into `h` if printed.
Delete ` method="get"` (it's the default).

vrana commented on adminer/processlist.inc.php:
```diff
@@ -51,6 +51,24 @@
+<?php } else { ?>
+<form action="<?php echo ME; ?>" method="get">
+<?php hidden_fields_get(); ?>
+<input type="hidden" value="" name="processlist">
```
`name` then `value`.

djmattyg007 commented:
Auto-refresh is off when the page loads, so you have to manually turn it on. Presumably you're not trying to select rows while it's on because you've done that intentionally, so I don't think that's much of an issue.

It's very useful when you're trying to monitor what a database is doing (especially when it is under heavy load) as it saves you from having to manually refresh the page every few seconds.

djmattyg007 commented:
Thanks for the feedback on the code, I'll make the changes and update the PR soon.

vrana commented on adminer/include/design.inc.php:
```diff
@@ -22,6 +22,9 @@ function page_header($title, $error = "", $breadcrumb = array(), $title2 = "") {
 <meta name="robots" content="noindex">
 <meta name="referrer" content="origin-when-crossorigin">
+<?php if (!$error && isset($_GET["processlist"], $_GET["autorefresh"])) { ?>
+<meta http-equiv="refresh" content="<?php echo $_GET["autorefresh"]; ?>">
```
This has XSS. Be careful about escaping in cases like this.

vrana commented:
With quick auto-refresh, it would be very hard to kill a process. Does it make sense to keep the checked checkboxes checked (you'll need JavaScript to send the form on the background).

djmattyg007 commented:
If a process is long-running and causing problems, you're probably going to have enough time to hit the "Stop autorefresh" button and then check the checkbox. If you click "Stop autorefresh" and the process has finished, it doesn't matter anymore. Therefore, I don't think (at least at this stage) that keeping track of which checkboxes are checked is necessary right now. It will introduce additional complexity that isn't strictly required.

I'm just about to make the switch from a meta http-equiv tag to the Refresh header, and then I should be able to update this PR.

djmattyg007 commented:
I've rebased my branch on top of the latest commits on the master branch, squashed it all down to two commits, and made almost all of the changes you requested. The one change I haven't made is to switch to a `Refresh` HTTP header in the process list file.

~~While I think it would still be possible to use the `Refresh` header over the `meta http-equiv` tag, I don't believe it's reasonable to implement the refresh in any place other than the header (`adminer/include/design.inc.php`). This appears to be the only reliable place to determine whether or not an error occurred during the current request for any particular reason. While there is a global `$error` variable, it is not always filled out. Additionally...~~

While writing that, I realised that the reason you wanted it moved to the process list file is because that file would not be loaded unless there were no errors with the request, or there was an error but it did not prevent the actual process list from being rendered.

I'll make the necessary update now :+1:

djmattyg007 commented:
Okay, I've pushed up that last change. Let me know if anything else needs adjusting or improving.

vrana commented on adminer/processlist.inc.php:
```diff
@@ -51,6 +51,25 @@
 </form>
+<?php if (isset($_GET["autorefresh"]) && !$error) { ?>
+<a href="<?php echo remove_from_uri("autorefresh"); ?>"><?php echo lang('Stop auto refresh'); ?></a>
+<?php header("Refresh: " . intval($_GET["autorefresh"]));
```
This is too late, you need to send it before any HTML output.

vrana commented on adminer/processlist.inc.php:
```diff
@@ -51,6 +51,25 @@
 <input type="hidden" name="token" value="<?php echo $token; ?>">
 </form>
+<?php if (isset($_GET["autorefresh"]) && !$error) { ?>
+<a href="<?php echo remove_from_uri("autorefresh"); ?>"><?php echo lang('Stop auto refresh'); ?></a>
```
This is still not in `xx.inc.php`. You can add all new messages there by running `lang.php xx`.

djmattyg007 commented:
My apologies, this is what happens when I attempt to write code late at night. I've fixed up all the issues you reported.

vrana commented:
I've patched this in and fixed the coding style to match Adminer but then decided that I'm not interested in this change.

djmattyg007 commented:
Thank you for considering this proposal and taking the time to review it. It's much appreciated.
